crates/iroha_core/src/peers_gossiper.rs

```diff
@@ -0,0 +1,220 @@
+//! Peers gossiper is actor which is responsible for gossiping addresses of peers.
+//!
+//! E.g. peer A changes address, connects to peer B,
+//! and then peer B will broadcast address of peer A to other peers.
+
+use std::{
+    collections::{BTreeMap, BTreeSet},
+    time::Duration,
+};
+
+use iroha_config::parameters::actual::TrustedPeers;
+use iroha_data_model::peer::{Peer, PeerId};
+use iroha_futures::supervisor::{Child, OnShutdown, ShutdownSignal};
+use iroha_p2p::{Broadcast, UpdatePeers, UpdateTopology};
+use iroha_primitives::{addr::SocketAddr, unique_vec::UniqueVec};
+use iroha_version::{Decode, Encode};
+use parity_scale_codec::{Error, Input};
+use tokio::sync::mpsc;
+
+use crate::{IrohaNetwork, NetworkMessage};
+
+/// [`Gossiper`] actor handle.
+#[derive(Clone)]
+pub struct PeersGossiperHandle {
+    message_sender: mpsc::Sender<(PeersGossip, Peer)>,
+    update_topology_sender: mpsc::UnboundedSender<UpdateTopology>,
+}
+
+impl PeersGossiperHandle {
+    /// Send [`PeersGossip`] to actor
+    pub async fn gossip(&self, gossip: PeersGossip, peer: Peer) {
+        self.message_sender
+            .send((gossip, peer))
+            .await
+            .expect("Gossiper must handle messages until there is at least one handle to it")
+    }
+
+    /// Send [`UpdateTopology`] message on network actor.
+    pub fn update_topology(&self, topology: UpdateTopology) {
+        self.update_topology_sender
+            .send(topology)
+            .expect("Gossiper must accept messages until there is at least one handle to it")
+    }
+}
+
+/// Actor which gossips peers addresses.
+pub struct PeersGossiper {
+    /// Peers provided at startup
+    initial_peers: BTreeMap<PeerId, SocketAddr>,
+    /// Peers received via gossiping from other peers
+    /// First-level key corresponds to `SocketAddr`
+    /// Second-level key - peer from which such `SocketAddr` was received
+    gossip_peers: BTreeMap<PeerId, BTreeMap<PeerId, SocketAddr>>,
```

Comment:
I think we need some eviction policy. This can be done separately

Reply:
Do you mean to periodically drop entries which were received long time ago?

Currently we have [eviction based on topology](https://github.com/dima74/iroha/blob/1194e61bd0aee33e259e210deefde8be64770acc/crates/iroha_core/src/peers_gossiper.rs#L137-L144)

Reply:
> Currently we have [eviction based on topology](https://github.com/dima74/iroha/blob/1194e61bd0aee33e259e210deefde8be64770acc/crates/iroha_core/src/peers_gossiper.rs#L137-L144)

I see. Can a malicious actor overwhelm other peer's gossip handle?

Reply:
> Can a malicious actor overwhelm other peer's gossip handle?

Single malicious actor can't, since we [choose address using majority rule](https://github.com/dima74/iroha/blob/1194e61bd0aee33e259e210deefde8be64770acc/crates/iroha_core/src/peers_gossiper.rs#L193-L204)

Reply:
I get that, but a single malicious actor could send lots of different false addresses, couldn't it?

Reply:
Only last entry is kept in the map.

---

E.g. consider we have three good peers PeerA, PeerB and PeerC and one malicious PeerD. Let's observe how PeerA will determine address of some PeerX. PeerB and PeerC will send correct address `GoodPeerX.com:8000` to PeerA, and PeerD will send `Bad1PeerX.com:8000`. So `gossip_peers` for PeerA will look like this:
```
{
  "PeerX": {
    "PeerB": "GoodPeerX.com:8000",
    "PeerC": "GoodPeerX.com:8000",
    "PeerD": "Bad1PeerX.com:8000"
  }
}
```
Based on majority rule, PeerA will use `GoodPeerX.com:8000` as address for PeerX. Next, consider PeerD again sends another bad address `Bad2PeerX.com:8000` to PeerA. Now map will change to:
```
{
  "PeerX": {
    "PeerB": "GoodPeerX.com:8000",
    "PeerC": "GoodPeerX.com:8000",
    "PeerD": "Bad2PeerX.com:8000"
  }
}
```
Again using majority rule, `GoodPeerX.com:8000` will be selected.